includes/sanitizers/class-amp-style-sanitizer.php

```diff
@@ -349,230 +431,741 @@ private function process_link_element( DOMElement $element ) {
+	 *     @type string[] $property_blacklist          Disallowed properties.
+	 *     @type bool     $convert_width_to_max_width  Convert width to max-width.
+	 *     @type string   $stylesheet_url              Original URL for stylesheet when originating via link (or @import?).
+	 *     @type string   $stylesheet_path             Original filesystem path for stylesheet when originating via link (or @import?).
+	 * }
+	 * @return string Processed stylesheet.
+	 */
+	private function process_stylesheet( $stylesheet, $node, $options = array() ) {
+		$should_tree_shake = ! empty( $options['class_selector_tree_shaking'] );
+		unset( $options['class_selector_tree_shaking'] );
+		$cache_key = md5( $stylesheet . serialize( $options ) ); // phpcs:ignore WordPress.PHP.DiscouragedPHPFunctions.serialize_serialize
+
+		$cache_group = 'amp-parsed-stylesheet-v1';
+		if ( wp_using_ext_object_cache() ) {
+			$parsed = wp_cache_get( $cache_key, $cache_group );
+		} else {
+			$parsed = get_transient( $cache_key . $cache_group );
 		}
-		$stylesheet = preg_replace( '/overflow(-[xy])?\s*:\s*(auto|scroll)\s*;?\s*/', '', $stylesheet, -1, $overlow_count );
-		if ( $overlow_count > 0 && ! empty( $this->args['validation_error_callback'] ) ) {
-			call_user_func( $this->args['validation_error_callback'], array(
-				'code' => 'css_overflow_property_removed',
-				'node' => $element,
-			) );
+		if ( ! $parsed || ! isset( $parsed['stylesheet'] ) || ! is_array( $parsed['stylesheet'] ) ) {
+			$parsed = $this->parse_stylesheet( $stylesheet, $options );
+			if ( wp_using_ext_object_cache() ) {
+				wp_cache_set( $cache_key, $parsed, $cache_group );
+			} else {
+				// The expiration is to ensure transient doesn't stick around forever since no LRU flushing like with external object cache.
+				set_transient( $cache_key . $cache_group, $parsed, MONTH_IN_SECONDS );
+			}
+		}
+
+		if ( ! empty( $this->args['validation_error_callback'] ) && ! empty( $parsed['validation_errors'] ) ) {
+			foreach ( $parsed['validation_errors'] as $validation_error ) {
+				call_user_func( $this->args['validation_error_callback'], array_merge( $validation_error, compact( 'node' ) ) );
+			}
+		}
+
+		$stylesheet = '';
+		foreach ( $parsed['stylesheet'] as $stylesheet_part ) {
+			if ( is_array( $stylesheet_part ) ) {
+				list( $selectors_parsed, $declaration_block ) = $stylesheet_part;
+				if ( $should_tree_shake ) {
+					$selectors = array();
+					foreach ( $selectors_parsed as $selector => $class_names ) {
+						if ( 0 === count( array_diff( $class_names, $this->used_class_names ) ) ) { // If all class names are used in the doc.
+							$selectors[] = $selector;
+						}
+					}
+				} else {
+					$selectors = array_keys( $selectors_parsed );
+				}
+				if ( ! empty( $selectors ) ) {
+					$stylesheet .= implode( ',', $selectors ) . $declaration_block;
+				}
+			} else {
+				$stylesheet .= $stylesheet_part;
+			}
 		}
+
 		return $stylesheet;
```

Comment:
Would it help to call the `validation_error_callback` here if `$stylesheet` is empty and it's from a `<link>`? This would account for when a stylesheet has no rules that pertain to DOM classes. Like if a plugin enqueues a stylesheet for its shortcode on every request.

```
if ( empty( $stylesheet ) && ( 'link' === $node->nodeName ) && ! empty( $this->args['validation_error_callback'] ) ) {
        // Call validation callback
```
        
We'd probably have to deprecate [amp-default.css](https://github.com/Automattic/amp-wp/blob/develop/assets/css/amp-default.css) to do this.